src/ILCompiler.ReadyToRun/src/IBC/IBCProfileParser.cs

```diff
@@ -0,0 +1,785 @@
+                                        //                                        UTILS.Info(String.Format("Ignoring ParamMethodSpec due to non-standard owning type signature: 0x{0:x8}", blob.Token));
+                                    }
+                                }
+                            }
+                            else
+                            {
+                                // We have old V2 IBC data (ignore it)
+                                ignore = true;
+                            }
+                            break;
+
+                        case BlobType.ExternalNamespaceDef:
+                            //                            UTILS.CheckAndSetHighest(ref result.highestExternalNamespaceToken, blob.Token);
+                            break;
+
+                        case BlobType.ExternalTypeDef:
+                            //                            UTILS.CheckAndSetHighest(ref result.highestExternalTypeToken, blob.Token);
+                            break;
+
+                        case BlobType.ExternalSignatureDef:
+                            //                            UTILS.CheckAndSetHighest(ref result.highestExternalSignatureToken, blob.Token);
+                            break;
+
+                        case BlobType.ExternalMethodDef:
+                            //                            UTILS.CheckAndSetHighest(ref result.highestExternalMethodToken, blob.Token);
+                            break;
+
+                        case BlobType.MetadataStringPool:
+                        case BlobType.MetadataGuidPool:
+                        case BlobType.MetadataBlobPool:
+                        case BlobType.MetadataUserStringPool:
+                            // These blob types should not be carried forward.
+                            ignore = true;
+                            break;
+                    }
+
+                    if (!ignore)
+                    {
+                        blobs.Add(new IBCBlobKey(blob.Token, blob.Type), blob);
+                    }
+                }
+            }
+            return blobs;
+        }
+
+
+        /// <summary>
+        /// Initialize the moduleId and typeToken
+        /// </summary>
+        /// <returns>false on failure</returns>
+        public static bool InitializeModuleIdAndTypeToken(out ModuleId moduleId, out uint typeToken, BlobReader typeSig)
+        {
+
+            moduleId = ModuleId.CurrentModule;
+            typeToken = 0;
+
+            byte currentTypeSigByte = typeSig.ReadByte();
+
+            if (currentTypeSigByte == (byte)CorElementType.ELEMENT_TYPE_MODULE_ZAPSIG)
+            {
+                // Get the external module index
+                uint index = (uint)typeSig.ReadCompressedInteger();
+                if (index == 0)
+                    return false;
+
+                // Get the new moduleId
+                moduleId = (ModuleId)(CorTokenType.mdtAssemblyRef + index);
```

Comment:
It might be worth calling out in a comment that this won't generally work in the large version bubble case where there are two different encoding for external modulerefs, one using the mdtAssemblyRef table and the other using the assemblyref table in the manifest metadata so that this code will need fixing to also support the latter variant.

Reply:
There's actually a bigger problem here. The code that generates the IBC data doesn't actually use any assemblyref other than the ones present in the initial image. The end result is that this logic is unable to reference modules that were not initially referenced by the application. This wasn't so bad on desktop framework, but with CoreCLR as many implementations are actually referenced through a type forwarder, there are significant problems.

Reply:
So, I'll add a comment, but the problem remains. (The current implementation of IBC emission from the runtime will emit data that always works.)